Cloud storage shutdown hang

andrwng opened the pull request:
This PR cleans up shutdown in a few places that, in summation contributed to a hang in shutdown when tiered storage is enabled. This also adds some logs, and updates some methods to coroutines to facilitate logging.

Fixes #8331

<!--

See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md##pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED.

Describe, in plain language, the motivation behind the change (bug fix,
feature, improvement) in this PR and how the included commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.

  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.

  Backport of PR #PR-NUMBER

-->

## Backports Required

<!--

Checking at least one of the checkboxes is REQUIRED if this PR is not a backport.

-->

- [ ] none - not a bug fix
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v22.3.x
- [ ] v22.2.x
- [ ] v22.1.x

## UX Changes

<!--

Content in this section is OPTIONAL.

Describe, in plain language, how this PR affects an end-user. Explain
topic flags, configuration flags, command line flags, deprecation
policies, etc. that are added or modified. Don't ship user breaking
changes. Ask the @redpanda-data/product team if you need help with user
visible changes.

-->

## Release Notes

### Bug Fixes

* Fixed a hang at shutdown when using tiered storage.

<!--

Adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

  ### Bug Fixes

  * Short description of the bug fix if this is a PR to `dev` branch.

  ### Features

  * Short description of the feature. Explain how to configure.

  ### Improvements

  * Short description of how this PR improves existing behavior.

If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

  * none

-->

jcsp commented on src/v/cloud_storage/tests/remote_partition_test.cc:
```diff
@@ -2178,6 +2179,63 @@ FIXTURE_TEST(
+    ss::maybe_yield().get();
+    ss::sleep(std::chrono::milliseconds(10)).get();
+    api.shutdown_connections();
+    g.close().get();
```
When this fails, does it manifest as a hang of the test?  If so, that was fine as a short term reproducer, but for committing it we should wrap this in a spin_wait_with_timeout or similar, so that the test fails cleanly.

andrwng replied:
Done

jcsp commented:
The other fanout fix PR merged, so this can be rebased + drop the commit on that file.

jcsp commented on src/v/storage/parser.cc:
```diff
@@ -217,22 +217,22 @@ static ss::future<result<model::record_batch_header>> read_header_impl(
 ss::future<result<model::record_batch_header>>
 continuous_batch_parser::read_header() {
-    return read_header_impl(get_stream(), *_consumer, _recovery);
+    auto& st = get_stream();
```
nit: this one was arguably easier to read before: is this perhaps a place that you coroutinized to add some debug, but it's gone now?

andrwng replied:
Yep, this is leftover from some debugging. I'll actually move the coroutinizing into a separate PR, to keep this one focused on actual fixes.

jcsp commented on src/v/cloud_storage/remote.cc:
```diff
@@ -100,6 +102,7 @@ ss::future<> remote::stop() {
 void remote::shutdown_connections() {
     cst_log.debug("Shutting down remote connections...");
     _pool.shutdown_connections();
+    _as.request_abort();
```
I think this is fine, but it shouldn't be functionally necessary, right?  If it was, then I would think something was wrong elsewhere.

If this is just to help with logging, then it's fine, but maybe add a comment to explain that.

andrwng replied:
Yeah, this was one of the first things I was suspicious of, but it ended up not being necessary. I'll remove it from the PR.

jcsp commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -304,13 +303,13 @@ class partition_record_batch_reader_impl final
 
-        // The reader may have been left in an indeterminate state.
-        // Re-set the pointer to it to ensure that it will not be reused.
+        // Regardless of which error, the reader may have been left in an
```
nit: this confused me for a moment, until I realized that we only fall through here in case of errors: maybe tweak the comment to mention that this path is only taken in case of exceptions from the above try{}

andrwng replied:
Done

andrwng commented:
I removed the rejiggering of the `remote` `abort_source`, and moved the coroutinizing commit into a separate PR in hopes that'll make this one a bit more backportable https://github.com/redpanda-data/redpanda/pull/8653

jcsp commented:
Failures were https://github.com/redpanda-data/redpanda/issues/8662

vbotbuildovich commented:
Failed to run cherry-pick command. I executed the below command:

```
git cherry-pick -x f72a7ed8f6e0ed825542b59e93f8cdca3a80a2fb b9cf31f7276a6f591b952b37d0315c9bf4f5ab3a d7e3f19bbbdc0afab3a02ebe9daff3494fad9bc6 9488d4ad2a7d17cfc32450af59db4b7ed5605f27 fe657df8ff084a83d3b1211df589191738388842 4c133a0801d11223463cffb44ed92a6aeecd8d99 bb6ef0409ad0f908cd9a0ca987343975e01dab96
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/4125837211).

dotnwat reviewed (commented):
great set of changes!
